File: docs/article/source/scala/apps/cargo-tracker/app/cargotracker/routing/domain/model/valueobjects/VoyageNumber.scala
package cargotracker.routing.domain.model.valueobjects

/** 航海の業務キー識別子（domain-model.md 準拠）。 形式は運送会社プレフィックス + 数字 / 英数字。例: `VY-001`、`MAEU-12345`。
  *
  * IT2 では緩めの正規表現で受け付ける。 IT3 以降に運送会社別の厳密な形式を導入する。
  */
opaque type VoyageNumber = String

object VoyageNumber:
  sealed trait Error
  case object EmptyValue extends Error
  case object InvalidFormat extends Error

  private val Pattern = "^[A-Z][A-Z0-9-]{1,19}$".r

  def apply(raw: String): Either[Error, VoyageNumber] =
    Option(raw) match
      case None | Some("") => Left(EmptyValue)
      case Some(v) =>
        if Pattern.matches(v) then Right(v) else Left(InvalidFormat)

  def unsafeFrom(raw: String): VoyageNumber = raw

  extension (vn: VoyageNumber) def value: String = vn

File: docs/article/source/java-4/apps/cargo-tracker/backend/routingms/src/main/java/com/example/cargotracker/routing/domain/model/valueobjects/VoyageNumber.java
package com.example.cargotracker.routing.domain.model.valueobjects;

import com.example.cargotracker.shared.domain.error.BusinessRuleViolation;

/**
 * 航海番号（Routing の識別子）。
 *
 * <p><b>BC ごとに別の型</b>にする（domain-model.md）。Handling も航海番号を持つが、
 * 意味も検査も違うので共有カーネルには置かない。</p>
 */
public record VoyageNumber(String value) {

    public VoyageNumber {
        if (value == null || value.isBlank()) {
            throw new BusinessRuleViolation("航海番号は必須です");
        }
        if (value.length() > 20) {
            throw new BusinessRuleViolation("航海番号は 20 文字以内です: " + value);
        }
    }
}
